CMakeLists.txt

```diff
@@ -99,20 +99,13 @@ if(BUILD_VERSION)
   )
 endif()
 
-# We need to set an optimization level.
-# Default to -Os
-# unless CONFIG_NO_OPTIMIZATIONS is set, then it is -O0
-# or unless CONFIG_DEBUG is set, then it is -Og
-#
-# also, some toolchain's break with -Os, and some toolchain's break
-# with -Og so allow them to override what flag to use
-#
-# Finally, the user can use Kconfig to add compiler options that will
-# come after these options and override them
-set_ifndef(OPTIMIZE_FOR_NO_OPTIMIZATIONS_FLAG "-O0")
-set_ifndef(OPTIMIZE_FOR_DEBUG_FLAG            "-Og")
-set_ifndef(OPTIMIZE_FOR_SIZE_FLAG             "-Os")
-set_ifndef(OPTIMIZE_FOR_SPEED_FLAG            "-O2")
+# @Intent: Set compiler specific flags related to optimizations
+# toolchain_cc_optimizations() macro must set:
+#   OPTIMIZE_FOR_NO_OPTIMIZATIONS_FLAG
+#   OPTIMIZE_FOR_DEBUG_FLAG
+#   OPTIMIZE_FOR_SIZE_FLAG
+#   OPTIMIZE_FOR_SPEED_FLAG
+toolchain_cc_optimizations()
```

Comment:
We need to do something about the inconsistency between toolchain_cc_optimizations(), and toolchain_cc_security_fortify().

One modifies the build environment, the other sets variables (that could be used to modify the build environment).

Because macros with side-effects should be avoided I would recommend resolving this inconsistency by rewriting it to look like this:

toolchain_cc_optimize_for_debug_flag(OPTIMIZE_FOR_DEBUG_FLAG)
toolchain_cc_optimize_for_speed_flag(OPTIMIZE_FOR_SPEED_FLAG)
toolchain_cc_optimize_for_no_optimizations_flag(OPTIMIZE_FOR_NO_OPTIMIZATIONS_FLAG)

Then the prototypes are different, and it is therefore clear that they will also behave different.

Reply:
I'd say that as long as the expectations have been properly documented (as comments) in root `CMakeLists.txt` the inconsistency is acceptable considering the gain in actually having abstracted the toolchain usage.fa
Since much of the current CMake logic is already side-effecting, these patches only factor that out. Non-functional changes. 

So I'd argue for first getting the value in (abstract toolchains), and then fixing cosmetics.

Reply:
> I'd say that as long as the expectations have been properly documented (as comments) in root CMakeLists.txt the inconsistency is acceptable

I disagree. When there is a choice between having unreadable code+comments, versus just readable code, we must choose readable code.

Reply:
why dont we move the related sections setting the optimization into the macro?

```
if(CONFIG_NO_OPTIMIZATIONS)
  set(OPTIMIZATION_FLAG ${OPTIMIZE_FOR_NO_OPTIMIZATIONS_FLAG})
elseif(CONFIG_DEBUG_OPTIMIZATIONS)
  set(OPTIMIZATION_FLAG ${OPTIMIZE_FOR_DEBUG_FLAG})
elseif(CONFIG_SPEED_OPTIMIZATIONS)
  set(OPTIMIZATION_FLAG ${OPTIMIZE_FOR_SPEED_FLAG})
elseif(CONFIG_SIZE_OPTIMIZATIONS)
  set(OPTIMIZATION_FLAG ${OPTIMIZE_FOR_SIZE_FLAG}) # Default
else()
  assert(0 "Unreachable code. Expected optimization level to have been chosen. See Kconfig.zephyr")
endif()

zephyr_compile_options(${OPTIMIZATION_FLAG})

```

Then it will be consistent.

`toolchain_cc_optimize_for_no_optimizations_flag` is funny macro name. if we go this way, then maybe `toolchain_cc_set_no_optimizations_flag`

Reply:
> Because unlike the other toolchain_cc_*(void) functions, this function will modify the build environment, and _also_ set a global variable that is used elsewhere.

I fail to see the difference, both modify the build environment.

Reply:
I agree with @nashif.

> As opposed to having toolchain_cc_*(void) consistently change the build environment.

IMHO it is OK for a toolchain_cc-macro to apply the compile flags.

Reply:
> I fail to see the difference, both modify the build environment.

One modifies a global variable and the other does not.